packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,257 @@
+import cdk = require('@aws-cdk/cdk');
+import { PolicyStatement } from "./policy-document";
+import { IPrincipal } from "./principals";
+
+/**
+ * Basic options for a grant operation
+ */
+export interface CommonGrantOptions {
+  /**
+   * The principal to grant to
+   *
+   * @default No work is done
+   */
+  principal?: IPrincipal;
+
+  /**
+   * The actions to grant
+   */
+  actions: string[];
+
+  /**
+   * The resource ARNs to grant to
+   */
+  resourceArns: string[];
+}
+
+/**
+ * Options for a grant operation
+ */
+export interface GrantWithResourceOptions extends CommonGrantOptions {
+  /**
+   * The resource with a resource policy
+   *
+   * The statement will be added to the resource policy if it couldn't be
+   * added to the principal policy.
+   */
+  resource: IResourceWithPolicy;
+
+  /**
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+/**
+ * Options for a grant operation that only applies to principals
+ */
+export interface GrantOnPrincipalOptions extends CommonGrantOptions {
+  /**
+   * Construct to report warnings on in case grant could not be registered
+   */
+  scope?: cdk.IConstruct;
+}
+
+/**
+ * Options for a grant operation to both identity and resource
+ */
+export interface GrantOnPrincipalAndResourceOptions extends CommonGrantOptions {
+  /**
+   * The resource with a resource policy
+   *
+   * The statement will always be added to the resource policy.
+   */
+  resource: IResourceWithPolicy;
+
+  /**
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+}
+
+export class Permissions {
+  /**
+   * Grant the given permissions to the principal
+   *
+   * The permissions will be added to the principal policy primarily, falling
+   * back to the resource policy if necessary. The permissions must be granted
+   * somewhere.
+   *
+   * - Trying to grant permissions to a principal that does not admit adding to
+   *   the principal policy while not providing a resource with a resource policy
+   *   is an error.
+   * - Trying to grant permissions to an absent principal (possible in the
+   *   case of imported resources) leads to a warning being added to the
+   *   resource construct.
+   */
+  public static grantWithResource(options: GrantWithResourceOptions): Grant {
```

Comment:
I would make all of these static methods of `Grant`. Would make a much nicer API:

```ts
return Grant.withResource(...)
```

Reply:
You also return a grant, so those are proper factor methods

Reply:
You're right, that's much better.